Library/Homebrew/test/test_formula.rb

```diff
@@ -406,3 +428,139 @@ def test_pour_bottle_dsl
+    tab
+  end
+
+  def rewrite_tab_with_tap(tab, tap_string)
+    tab.source["tap"] = tap_string
+    tab.tabfile.unlink if tab.tabfile.exist?
+    tab.write
+  end
+
+  def test_greater_different_tap_installed
+    greater_prefix.mkpath
+    greater_tab = tab_for_prefix(greater_prefix, "user/repo")
+    greater_tab.write
+
+    assert_predicate f.outdated_versions, :empty?
+  end
+
+  def test_greater_same_tap_installed
+    f.instance_variable_set(:@tap, CoreTap.instance)
+
+    greater_prefix.mkpath
+    greater_tab = tab_for_prefix(greater_prefix, "homebrew/core")
+    greater_tab.write
+
+    assert_predicate f.outdated_versions, :empty?
+  end
+
+  def test_outdated_different_tap_installed
+    outdated_prefix.mkpath
+    outdated_tab = tab_for_prefix(outdated_prefix, "user/repo")
+    outdated_tab.write
+
+    refute_predicate f.outdated_versions, :empty?
+  end
+
+  def test_outdated_same_tap_installed
+    f.instance_variable_set(:@tap, CoreTap.instance)
+
+    outdated_prefix.mkpath
+    outdated_tab = tab_for_prefix(outdated_prefix, "homebrew/core")
+    outdated_tab.write
+
+    refute_predicate f.outdated_versions, :empty?
+  end
+
+  def test_same_head_installed
+    f.instance_variable_set(:@tap, CoreTap.instance)
+
+    head_prefix.mkpath
+    head_tab = tab_for_prefix(head_prefix, "homebrew/core")
+    head_tab.write
+
+    assert_predicate f.outdated_versions, :empty?
+  end
+
+  def test_different_head_installed
+    f.instance_variable_set(:@tap, CoreTap.instance)
+
+    head_prefix.mkpath
+    head_tab = tab_for_prefix(head_prefix, "user/repo")
+    head_tab.write
+
+    assert_predicate f.outdated_versions, :empty?
+  end
+
+  def test_mixed_taps_greater_version_installed
+    f.instance_variable_set(:@tap, CoreTap.instance)
+
+    outdated_prefix.mkpath
+    greater_prefix.mkpath
+    outdated_tab = tab_for_prefix(outdated_prefix, "homebrew/core")
+    greater_tab = tab_for_prefix(greater_prefix, "user/repo")
+    outdated_tab.write
+    greater_tab.write
+
+    assert_predicate f.outdated_versions, :empty?
+
+    rewrite_tab_with_tap(greater_tab, "homebrew/core")
+
+    assert_predicate f.outdated_versions, :empty?
+  end
+
+  def test_mixed_taps_outdated_version_installed
+    f.instance_variable_set(:@tap, CoreTap.instance)
+
+    extra_outdated_prefix = HOMEBREW_CELLAR/"#{f.name}/1.0"
+    extra_outdated_prefix.mkpath
+    outdated_prefix.mkpath
+
+    outdated_tab = tab_for_prefix(outdated_prefix)
+    extra_outdated_tab = tab_for_prefix(extra_outdated_prefix, "homebrew/core")
+    extra_outdated_tab.write
+    outdated_tab.write
+
+    refute_predicate f.outdated_versions, :empty?
+
+    rewrite_tab_with_tap(outdated_tab, "user/repo")
+
+    refute_predicate f.outdated_versions, :empty?
+  end
+
+  def test_same_version_tap_installed
+    f.instance_variable_set(:@tap, CoreTap.instance)
+
+    same_prefix.mkpath
+    same_tab = tab_for_prefix(same_prefix, "homebrew/core")
+    same_tab.write
```

Comment:
:+1: I don't remove `tab` at the end of the method, do I need to? Now it's not used, that is true, but I'm not sure if I need to remove it.